keps/sig-network/3015-node-level-topology/README.md

```diff
@@ -0,0 +1,458 @@
+  Endpoints] should solve the problems that made
+  `externalTrafficPolicy: Local` unreliable for some cases.
+
+- Attempting to improve the endpoint selection algorithm for services
+  with a number of endpoints that is not (roughly) an integer multiple
+  of the number of nodes. If you have a service with 15 endpoints in a
+  cluster with 10 nodes, then the service could potentially be
+  optimized by making clients more likely to select a local endpoint,
+  but they can't _exclusively_ select local endpoints or the endpoints
+  on nodes with only a single endpoint would presumably get
+  overloaded. But this KEP does not attempt to figure out any new
+  behavior there.
+
+[Proxy Terminating Endpoints]: ../1669-graceful-termination-local-external-traffic-policy/
+
+## Proposal
+
+### User Stories
+
+#### Per-Node DNS Pods
+
+As a cluster administrator, I plan to run a DNS pod on each node, and
+would like DNS requests from other pods to always go to the local DNS
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should automatically just go to a remote pod instead.
+
+(This is a modified version of a user story from the [Internal Traffic
+Policy KEP]. The original implies (by omission) that is is acceptable
+to have DNS lookups fail when there is no local DNS pod.)
+
+[Internal Traffic Policy KEP]: ../2086-service-internal-traffic-policy/
+
+## Design Details
+
+### Applying Node-Level Topology
+
+Given a Service to which node-level topology applies (see next
+section), the implementation is simple: if there is at least one
+endpoint for the service on the local node, then filter the list of
+endpoints to include only local ones. (If there are no local
+endpoints, then use the same rules as would have been used if
+node-level topology was not in effect.)
+
+As with Topology Aware Hints, Node-Level Topology would only apply to
+connections with `Cluster` traffic policy, because
+`internalTrafficPolicy: Local` semantically _requires_ local delivery
+(and is not allowed to fall back when there are no local endpoints),
+and `externalTrafficPolicy: Local` requires source IP preservation
+(which is generally not possible when sending to a non-local
+endpoint).
+
+### Choosing Services to Apply Node-Level Topology To
+
+```
+<<[UNRESOLVED]>>
+
+Pick an approach.
+
+The Automatic approach seems simplest (and would have the effect of
+immediately making DNS faster in most(?) clusters). But maybe it would
+do bad things in some circumstances?
+
+<<[/UNRESOLVED]>>
+```
+
+#### Approach 1: Manual
+
+One approach would be to allow users to explicitly tag services where
+they want this behavior.
+
+This could be implemented via the existing
+`"service.kubernetes.io/topology-aware-hints"` annotation perhaps.
+Instead of setting it to `Auto`, the user could set it to `Node` to
+indicate node-level topology rather than zone-level.
+
+(This would imply that a service could not use both node-level
```

Comment:
Or set it to "NodeZone" ?

Reply:
Actually, isn't it important that both be possible, so version-skewed (old) proxies which don't know this new hint would fall back on regular topology, and REALLY old ones would fall back on all slices?

Reply:
~3.5-4 years ago we talked about this being an array `["node", "zone", "*"]` or whatever, enabling the user to state their preferences. We abandoned that because we decided fully automatic is much better, I guess we're circling back around? Or I guess I see automatic mode below.